L1Trigger/RPCTriggerPrimitives/plugins/PrimitiveAlgoFactory.cc

```diff
@@ -0,0 +1,4 @@
+#include "FWCore/PluginManager/interface/PluginFactory.h"
+#include "L1Trigger/RPCTriggerPrimitives/plugins/PrimitiveAlgoFactory.h"
+
+EDM_REGISTER_PLUGINFACTORY(PrimitiveAlgoFactory, "PrimitiveAlgoFactory");
```

Comment:
This is a big problem. The exact same Factory is already registered under a different name:
https://github.com/cms-sw/cmssw/blob/04a85943355965447837cff7da77bde47a88e176/RecoLocalMuon/RPCRecHit/interface/RPCRecHitAlgoFactory.h#L13

One cannot registered two factories with identically signatures under two different names. Only one name will be used by the system which leads to problems loading the plugins. 

This factory declaration should go away and the code should just use the existing RPCRecHitAlgoFactory.

Reply:
Unfortunately, I have problems by calling the rechitalgofactory, Does anyone know how can I solve?
L1TMuonRPCTriggerPrimitivesProducer.cc:(.text+0x28e): undefined reference to `edmplugin::PluginFactory<RPCRecHitBaseAlgo* (edm::ParameterSet const&)>::get()'

Reply:
Do you have
```
<use name="RecoLocalMuon/RPCRecHit">
```
to your BuildFile.xml?

Reply:
Yes I do, but it does not work.

Reply:
The problem is the package RecoLocalMuon/RPCRecHit creates a plugin not a library and a plugin. Therefore you can't link to it and get the symbol you need. So it looks like files within RecoLocalMuon/RPCRecHit needs to be moved around so that at least RecoLocalMuon/RPCRecHit/src/RPCRecHitAlgoFactory.cc gets compiled into a library which is shared between the two packages.

Reply:
"RecoLocalMuon/RPCRecHit/interface/RPCRecHitBaseAlgo.h" would also have to be in the shared library.

Reply:
I copied the two files into the local plugin folder, and now it is working but I don't know if this is correct.

Reply:
No, that will just lead to the problem we had before.

Reply:
I ran some tests in raw data and it is OK

Reply:
RecoLocalMuon/RPCRecHit needs to be re-configured
- Add a plugins directory
- Move the following files to the pugins directory
   - all files in interface/ except RPCRecHitAlgoFactory.h, RPCRecHitBaseAlgo.h
   - all files in src/ except RPCRecHitAlgoFactory.cc, RPCRecHitBaseAlgo.cc, RPCRollMask.h
- Copy BuildFile.xml to plugins/BuildFile.xml and add `<use name="RecoLocalMuon/RPCRecHit" />`
- Remove `<flags EDM_PLUGIN="1"/>` from the top level BuildFile.xml file.

Reply:
Hi Thanks, 
I got the following error by doing those changes:

tmp/slc7_amd64_gcc700/src/RecoLocalMuon/RPCRecHit/src/RecoLocalMuonRPCRecHit/RPCRecHitStandardAlgo.cc.o:(.data.rel.ro._ZTI21RPCRecHitStandardAlgo[_ZTI21RPCRecHitStandardAlgo]+0x10): undefined reference to `typeinfo for RPCRecHitBaseAlgo'
tmp/slc7_amd64_gcc700/src/RecoLocalMuon/RPCRecHit/src/RecoLocalMuonRPCRecHit/RPCRecHitStandardAlgo.cc.o:(.data.rel.ro._ZTV21RPCRecHitStandardAlgo[_ZTV21RPCRecHitStandardAlgo]+0x28): undefined reference to `RPCRecHitBaseAlgo::reconstruct(RPCRoll const&, RPCDetId const&, std::pair<__gnu_cxx::__normal_iterator<RPCDigi const*, std::vector<RPCDigi, std::allocator<RPCDigi> > >, __gnu_cxx::__normal_iterator<RPCDigi const*, std::vector<RPCDigi, std::allocator<RPCDigi> > > > const&, std::bitset<192ul> const&)'
collect2: error: ld returned 1 exit status

Reply:
those changes generate similar errors. 
tmp/slc7_amd64_gcc700/src/RecoLocalMuon/RPCRecHit/src/RecoLocalMuonRPCRecHit/RPCRecHitBaseAlgo.cc.o: In function `RPCRecHitBaseAlgo::reconstruct(RPCRoll const&, RPCDetId const&, std::pair<__gnu_cxx::__normal_iterator<RPCDigi const*, std::vector<RPCDigi, std::allocator<RPCDigi> > >, __gnu_cxx::__normal_iterator<RPCDigi const*, std::vector<RPCDigi, std::allocator<RPCDigi> > > > const&, std::bitset<192ul> const&)':
RPCRecHitBaseAlgo.cc:(.text+0x71): undefined reference to `RPCClusterizer::doAction(std::pair<__gnu_cxx::__normal_iterator<RPCDigi const*, std::vector<RPCDigi, std::allocator<RPCDigi> > >, __gnu_cxx::__normal_iterator<RPCDigi const*, std::vector<RPCDigi, std::allocator<RPCDigi> > > > const&)'

Reply:
Wasn't RPCRecHitStandardAlgo.cc one of the files to move out of src/  and into plugins/ ?

Reply:
Indeed, StandardAlgo is now in plugins

Reply:
Can you point us to your code?

@smuzaffar any ideas?

Reply:
/afs/cern.ch/user/m/masegura/CMSSW_10_6_0/src/RecoLocalMuon/RPCRecHit